cktap/proto.py

```diff
@@ -316,8 +316,52 @@ def get_slot_usage(self, slot, cvc=None):
 
         return (addr, status, here)
 
+    def sign_digest(self, cvc: str, slot: int, digest: bytes, subpath: str = None) -> bytes:
+        """
+        Sign 32 bytes digest and return 65 bytes long recoverable signature.
+
+        Uses derivation path based on current set derivation on card plus optional
+        subpath parameter which if provided, will be added to card derivation path.
+        Subpath can only be of length 2 and non-hardened components only.
+
+        Returns non-deterministic recoverable signature (header[1b], r[32b], s[32b])
+        """
+        if len(digest) != 32:
+            raise ValueError("Digest must be exactly 32 bytes")
+        if not self.is_tapsigner and subpath:
+            raise ValueError("Cannot use 'subpath' option for SATSCARD")
+        # subpath validation
+        int_path = str2path(subpath) if subpath is not None else []
+        if len(int_path) > 2:
+            raise ValueError(f"Length of path {subpath} greater than 2")
+        if not none_hardened(int_path):
+            raise ValueError(f"Subpath {subpath} contains hardened components")
+        if self.is_tapsigner:
+            slot = 0
+        for _ in range(5):
+            try:
+                if self.is_tapsigner:
+                    ses_key, resp = self.send_auth('sign', cvc, slot=slot, digest=digest, subpath=int_path)
+                else:
+                    ses_key, resp = self.send_auth('sign', cvc, slot=slot, digest=digest)
+                expect_pub = resp['pubkey']
+                sig = resp['sig']
+                if not CT_sig_verify(expect_pub, digest, sig):
+                    # this is probably not needed
```

Comment:
this is how core does it iirc

Reply:
keep it IMHO; not a performance concern, and good extra check